### ISSUE TITLE
Add source filter configuration to @Query repository methods

### ISSUE THREAD

**Issue by nathan-t-lr** [spring-projects/spring-data-elasticsearch#2146]
The `Query` annotation is very limited in that it can only search using matchers but not specify additional configuration like source filters or aggregations. 

Because of this, if advanced searching is desired we must interact directly with `ElasticsearchRestTemplate` using `NativeSearchQuery`s or other means to get the results we want.

We should implement an additional parameter in `@Query` and parse these configurations to add them to the search query. This will be much more convenient and require much less setup time for elasticsearch repository searches.
 
Closes #2062

**Comment by sothawo** [comment]
Please separate the two parts: source filter and aggregations. They are different independent things. We should handle them separately in two issues with different pull requests. And let's first discuss the architecture and how to solve it  before implementing.

Thinking about it, it might be a better solution not to add this to the `@Query` annotation. Source filters and aggs are not part of the query when sent to Elasticsearch. So putting them in a parameter `valueParams` is conceptually wrong, They are no values for query parameters as this name suggests. Value parameters for queries are added to the function and represented by the placeholders in the query string.

### source filters

I'd suggest to create a new `@SourceFilters`  annotation which has `includes` and `excludes` properties, both being String arrays. We already configure the highlighting with a separate annotation for example. The processing of this annotation would be according  what you had in the first PR.

### aggregations

This is more difficult. Whereas source filter can be handled on the `ElasticsearchStringQuery` and `ElasticsearchQueryMethod` level where there is no knowledge of the underlying client implementation, building the aggregations to be sent to Elasticsearch must be done in the client-specific implementation of the `ElasticsearchOperations` interface. So for this we must find a way to pass down the aggregations definitions (as a json string or in whatever form) to the point the the real request for Elasticsearch is built. 

But that's a topic for a different issue.

**Comment by nathan-t-lr** [comment]
> Please separate the two parts: source filter and aggregations. They are different independent things. We should handle them separately in two issues with different pull requests. And let's first discuss the architecture and how to solve it before implementing.
Might take me some time to separate it since the work week is starting but I will keep it in mind
 
> Thinking about it, it might be a better solution not to add this to the `@Query` annotation. Source filters and aggs are not part of the query when sent to Elasticsearch. So putting them in a parameter `valueParams` is conceptually wrong, They are no values for query parameters as this name suggests. Value parameters for queries are added to the function and represented by the placeholders in the query string.

I am thinking of `@Query` in the general spring sense, as in an annotation that lets you perform queries in whichever language. `value` in this sense only refers to the specific languages query input (take for example how hibernates implementation allows sql). This is probably what leads to the confusion about why whatever we put in `value` is wrapped in a query object. In this way, it is in my opinion appropriate to have source filter and aggregations on this annotation, but perhaps we can change the name of the argument from `valueParams` to something less ambiguous.

> ### aggregations
> This is more difficult. Whereas source filter can be handled on the `ElasticsearchStringQuery` and `ElasticsearchQueryMethod` level where there is no knowledge of the underlying client implementation, building the aggregations to be sent to Elasticsearch must be done in the client-specific implementation of the `ElasticsearchOperations` interface. So for this we must find a way to pass down the aggregations definitions (as a json string or in whatever form) to the point the the real request for Elasticsearch is built.

My implementation is working well it seems, no tests are broken and the aggregations are coming back. Take a look and let me know if there is any issue. If not I can implement it for the reactive code/anywhere else it needs to go. If you still think there is a problem maybe we can hop on a call and you can explain exactly the issue.

**Comment by sothawo** [comment]
I just had a short glance at your PR and left some remarks there, but a longer comment here:

### branches and versions

the development in Spring Data Elasticsearch is done on the _main_ branch for both new features and bug fixes. The _main_ branch is what will be the next released version (normally). Bug fixes are then backported by cherrypicking to the released branches. New features are only available in the next version.

Your PR is targeting 4.3.x, this is a released version where no new features will be added.

The current version is 4.3.x, the next version will be 4.4. And now we currently have a setup where _main_ is not the next branch. We have besides _main_  the _4.4.x_ branch, this will be the next release, planned for May 13. The reason for this is, that all Spring Data modules are in their _main_ branches already preparing for the big version jump in autumn With Spring 6, Spring Data 3, Spring Data Elasticsearch in 5 and all based on Java 17. And so in Spring Data Elasticsearch we are currently building the _main_ and _4.4.x_ in parallel. After the release of 4.4 work will go on as normal, but we might add new features to the then released version 4.4. This is related to the next topic, Elasticsearch client versions.

### Elasticsearch client versions

Up to version 4.4 the default client to communicate with Elasticsearch is the deprecated `RestHighLevelClient`. In 4.4 it will be possible to optionally use the new `ElasticsearchClient` instead of the `RestHighLevelClient`. Both of these clients are currently used with Elasticsearch version 7.17.3.

In Spring Data Elasticsearch 5, when switching to Elasticsearch 8 (what is planned) we don't have the `RestHighLevelClient` anymore, Elasticsearch does not have a version 8 for this. It might be possible to default to  the `ElasticsearchClient` 8 and optionally have the `RestHighLevelClient` 7.17, but this must be tested if possible (both use the low level `RestClient`  from Elasticsearch and we have to see if such a mixture can be done.

In 4.4 all code using the new `ElasticsearchClient` is in a package `org.springframework.data.elasticsearch.client.elc`. The code using the `RestHighLevelClient` is still in the packages where it is in 4.3, we could not move it in a minor version, because this would have been breaking changes, this will then be done in 5, if we can keep this code for one more version. But in the classes that are a level above that, that should not know which client is used under the hood, all references and imports of Elasticsearch classes are removed. So there is no point for example in having aggregation classes there.

In addition to that we might add support to use an Opensearch client to communicate with an OpenSearch cluster in the future.

### short remark on the aggregations

I saw that you have code to just build terms aggregations. I once checked the client libraries of Elasticsearch and found more than 80 classes to model different aggregations. So for Spring Data Elasticsearch we would need a way to have the Elasticsearch client library parse an aggregation json text, this is nothing we should do ourselves in Spring Data Elasticsearch.

### how to proceed

For the first thing I'd say, leave the aggs out . Let's first do the source filters. Have a look at the _main_ branch, the new classes and test setup that is there.

I'd still suggest a `@SourceFilter`  annotation with the `includes` and `excludes`. It would be nice to add this as a property to the `@Query` annotation, but Java does not allow for annotation properties to be null - which could be bypassed by having a default instance.

**Comment by nathan-t-lr** [comment]
Thanks for the feedback @sothawo, I was unaware that the branch I built against wasn't going to be released in the future. I will take the source filter code and merge it into the 4.4.x branch and save aggregations for later. 

The only trepidation I have with adding `@SourceFilter` is that the it sets up the style to be many annotations to add additional query config. If this is the case a user could go through the hassle of setting up an elasticsearch template I think. 
Reason I like it in the `@Query` is I usually test elasticsearch through postman, and it would be awesome if I could copy and past my queries into the `valueParam` attribute as it's defined on the pr i made. Let me know your thoughts.

At the end of the day, I'll implement it according to the style you guys ultimately want because these features would save devs a lot of time. Just want to make the new feature as convenient as possible :)

**Comment by nathan-t-lr** [comment]
Also for context, I implemented against 4.3.x because this is the version my team is using and I was using my own repo to test. But now I have got the hang of testing with the spring-data-elasticsearch tests so it shouldn't be a problem to switch to 4.4.x

**Comment by sothawo** [comment]
Please develop this against _main_! If this works, then we'll backport it to 4.4. As I said, 4.4 will live along 5.x for new features as well, atl least for a while. But development is always on the _main_ branch. And although _main_ is built with Java 17, because of this backport to Java 8, no things like text blocks (I'd like to use them).

**Comment by nathan-t-lr** [comment]
> Please develop this against main! If this works, then we'll backport it to 4.4.
Oh okay, `main` it is!!

**Comment by sothawo** [comment]
and yes, I know that this like a mess with the branches and client versions. But Elasticsearch brought out the new client within half a year so we had not much time to test and integrate it (and this is still an ongoing process) and in addition to that Spring prepares the next big change.

And when I write "we", the implementation of Spring Data Elasticsearch is a community project with basically almost only me working on it in my spare time, not during work, what vmware does is hosting all this stuff, doing the release plans and releases and we align on architectural topics, so that the different Spring Data modules are aligned.

**Comment by nathan-t-lr** [comment]
> And when I write "we", the implementation of Spring Data Elasticsearch is a community project with basically almost only me working on it in my spare time, not during work

Looks like I will be joining in on this project (similarly in my spare time off work), so hopefully we can get some good features out
